File: android/core/src/main/kotlin/com/openmausbot/companion/core/Endpoint.kt
package com.openmausbot.companion.core

import java.net.URI
import java.nio.charset.StandardCharsets
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

@Serializable
enum class CompanionEndpointKind {
    @SerialName("hosted")
    HOSTED,

    @SerialName("tailnet")
    TAILNET,

    @SerialName("lan")
    LAN,

    @SerialName("bonjour")
    BONJOUR,
}

/** Whether a route may carry a credential without another explicit local-network choice. */
enum class CompanionEndpointSecurityClass {
    PROTECTED,
    EXPLICIT_LOCAL,
}

/**
 * One validated, complete dialing authority advertised by the companion.
 *
 * Construction is intentionally fallible and the constructor is private: an endpoint cannot
 * exist without passing the scheme, authority and kind policy below. This keeps later pairing
 * and failover work from having to revalidate a value after it has already entered the domain.
 */
@ConsistentCopyVisibility
@Serializable(with = CompanionEndpointSerializer::class)
data class CompanionEndpoint private constructor(
    val url: String,
    val kind: CompanionEndpointKind,
    val priority: Int,
) {
    val baseUrl: URI
        get() = URI(url)

    val host: String
        get() {
            val parsed = baseUrl.host.orEmpty()
                // URI exposes an IPv6 zone in its encoded form. Connection's scoped-DNS path
                // needs the original '%' separator when it turns the host back into a socket.
                .replace("%25", "%", ignoreCase = true)
            return Connection.urlHost(parsed)
        }

    val port: Int
        get() = baseUrl.port.takeIf { it >= 0 } ?: if (isSecure) 443 else 80

    val isSecure: Boolean
        get() = baseUrl.scheme.equals("https", ignoreCase = true)

    val securityClass: CompanionEndpointSecurityClass
        get() = when (kind) {
            CompanionEndpointKind.HOSTED,
            CompanionEndpointKind.TAILNET,
            -> CompanionEndpointSecurityClass.PROTECTED

            CompanionEndpointKind.LAN,
            CompanionEndpointKind.BONJOUR,
            -> CompanionEndpointSecurityClass.EXPLICIT_LOCAL
        }

    val protectsCredentials: Boolean
        get() = securityClass == CompanionEndpointSecurityClass.PROTECTED

    /** Full HTTPS authority for hosted routes; familiar direct form for local routes. */
    val displayAddress: String
        get() = if (kind == CompanionEndpointKind.HOSTED || isSecure) {
            url
        } else if (port == DEFAULT_COMPANION_PORT) {
            host
        } else {
            "$host:$port"
        }

    companion object {
        const val MAX_URL_BYTES = 2_048
        const val MAX_PRIORITY = 1_000_000
        const val DEFAULT_COMPANION_PORT = 8810

        fun create(url: String, kind: CompanionEndpointKind, priority: Int): CompanionEndpoint? {
            if (priority !in 0..MAX_PRIORITY) return null
            val normalized = normalizedUrl(url, kind) ?: return null
            return CompanionEndpoint(normalized, kind, priority)
        }

        fun direct(
            host: String,
            port: Int,
            kind: CompanionEndpointKind? = null,
            priority: Int,
        ): CompanionEndpoint? {
            if (port !in 1..65_535) return null
            val normalizedHost = Connection.urlHost(host)
            if (normalizedHost.isEmpty()) return null
            val raw = runCatching {
                URI("http", null, normalizedHost, port, null, null, null).toASCIIString()
            }.getOrNull() ?: return null
            return create(raw, kind ?: inferredDirectKind(normalizedHost), priority)
        }

        fun inferredDirectKind(host: String): CompanionEndpointKind {
            val canonical = canonicalDnsHost(host)
            return when {
                validTailnetHost(canonical) -> CompanionEndpointKind.TAILNET
                canonical.endsWith(".local") -> CompanionEndpointKind.BONJOUR
                else -> CompanionEndpointKind.LAN
            }
        }

        /** Credential-safe candidates, in caller order, for the later pairing/failover passes. */
        fun automaticCandidates(candidates: List<CompanionEndpoint>): List<CompanionEndpoint> {
            val preferred = candidates.firstOrNull() ?: return emptyList()
            val seen = mutableSetOf<String>()
            return candidates.filter { candidate ->
                seen.add(candidate.url) && if (preferred.protectsCredentials) {
                    candidate.protectsCredentials
                } else {
                    candidate.url == preferred.url || candidate.protectsCredentials
                }
            }
        }

        private fun normalizedUrl(raw: String, kind: CompanionEndpointKind): String? {
            val trimmed = raw.trim()
            if (trimmed.toByteArray(StandardCharsets.UTF_8).size > MAX_URL_BYTES) return null

            val parsed = runCatching { URI(trimmed) }.getOrNull() ?: return null
            val scheme = parsed.scheme?.lowercase() ?: return null
            val host = parsed.host?.takeIf { it.isNotEmpty() } ?: return null
            if (!parsed.isAbsolute || parsed.rawAuthority == null) return null
            if (scheme != "http" && scheme != "https") return null
            if (parsed.rawUserInfo != null || parsed.rawQuery != null || parsed.rawFragment != null) return null
            if (parsed.rawPath.isNotEmpty() && parsed.rawPath != "/") return null
            if (parsed.port != -1 && parsed.port !in 1..65_535) return null
            // java.net.URI treats an empty explicit port as if no port was supplied.
            if (parsed.rawAuthority.endsWith(':')) return null

            when (kind) {
                CompanionEndpointKind.HOSTED -> if (scheme != "https") return null
                CompanionEndpointKind.TAILNET -> {
                    if (scheme != "http" || !validTailnetHost(canonicalDnsHost(host))) return null
                }
                CompanionEndpointKind.LAN,
                CompanionEndpointKind.BONJOUR,
                -> if (scheme != "http") return null
            }

            return runCatching {
                URI(scheme, null, host.lowercase(), parsed.port, null, null, null).toASCIIString()
            }.getOrNull()
        }

        private fun canonicalDnsHost(host: String): String =
            host.lowercase().removeSurrounding("[", "]").trimEnd('.')

        private fun validTailnetHost(host: String): Boolean =
            host.length > ".ts.net".length && host.endsWith(".ts.net")
    }
}

object CompanionEndpointSerializer : KSerializer<CompanionEndpoint> {
    @Serializable
    private data class Wire(
        val url: String,
        val kind: CompanionEndpointKind,
        val priority: Int,
    )

    override val descriptor: SerialDescriptor = Wire.serializer().descriptor

    override fun deserialize(decoder: Decoder): CompanionEndpoint {
        val wire = Wire.serializer().deserialize(decoder)
        return CompanionEndpoint.create(wire.url, wire.kind, wire.priority)
            ?: throw SerializationException(
                "Companion endpoints must be absolute authorities; hosted routes require " +
                    "HTTPS and tailnet routes require an HTTP .ts.net name.",
            )
    }

    override fun serialize(encoder: Encoder, value: CompanionEndpoint) {
        Wire.serializer().serialize(encoder, Wire(value.url, value.kind, value.priority))
    }
}

File: ios/Sources/CompanionCore/Endpoint.swift
import Foundation

/// Why an address exists. The kind is display and policy metadata; the URL
/// remains the complete dialing authority, so hosted HTTPS and local HTTP can
/// live in the same fallback list without guessing a scheme from a hostname.
public enum CompanionEndpointKind: String, Codable, CaseIterable, Sendable {
    case hosted
    case tailnet
    case lan
    case bonjour
}

/// The credential-handling boundary for a route.
///
/// Hosted HTTPS is authenticated by Web PKI. A Tailscale MagicDNS name is
/// authenticated and encrypted by the tailnet even though the local URL is
/// HTTP. LAN and Bonjour routes are deliberately cleartext and therefore
/// require an exact, explicit choice by the user; they are never generic
/// automatic fallbacks for a bearer token or a one-time pairing credential.
public enum CompanionEndpointSecurityClass: Sendable {
    case protected
    case explicitLocal
}

/// One validated route to the desktop companion.
public struct CompanionEndpoint: Codable, Hashable, Sendable {
    public let url: String
    public let kind: CompanionEndpointKind
    public let priority: Int

    public init?(url: String, kind: CompanionEndpointKind, priority: Int) {
        guard (0...1_000_000).contains(priority),
              let normalized = Self.normalizedURL(url, kind: kind)
        else { return nil }
        self.url = normalized
        self.kind = kind
        self.priority = priority
    }

    public var baseURL: URL? { URL(string: url) }

    public var host: String {
        guard let host = URLComponents(string: url)?.host else { return "" }
        return Connection.urlHost(host)
    }

    public var port: Int {
        guard let components = URLComponents(string: url) else { return 0 }
        return components.port ?? (components.scheme?.lowercased() == "https" ? 443 : 80)
    }

    public var isSecure: Bool {
        URLComponents(string: url)?.scheme?.lowercased() == "https"
    }

    public var securityClass: CompanionEndpointSecurityClass {
        switch kind {
        case .hosted, .tailnet: return .protected
        case .lan, .bonjour: return .explicitLocal
        }
    }

    public var protectsCredentials: Bool { securityClass == .protected }

    /// Host-only for the old direct routes, full HTTPS authority for hosted
    /// routes. Used in status copy, never for dialing.
    public var displayAddress: String {
        if kind == .hosted || isSecure { return url }
        return port == 8810 ? host : "\(host):\(port)"
    }

    /// Construct the legacy HTTP route represented by `host` + `port`.
    public static func direct(
        host: String,
        port: Int,
        kind: CompanionEndpointKind? = nil,
        priority: Int
    ) -> CompanionEndpoint? {
        let resolvedKind = kind ?? inferredDirectKind(host)
        var components = URLComponents()
        components.scheme = "http"
        components.host = Connection.urlHost(host)
        components.port = port
        guard let value = components.url?.absoluteString else { return nil }
        return CompanionEndpoint(url: value, kind: resolvedKind, priority: priority)
    }

    /// Older saved connections only carry host strings. Recover the security
    /// class from names whose ownership has a useful transport meaning rather
    /// than treating a protected Tailscale name as arbitrary LAN cleartext.
    public static func inferredDirectKind(_ host: String) -> CompanionEndpointKind {
        let canonical = canonicalDNSHost(host)
        if validTailnetHost(canonical) { return .tailnet }
        if canonical.hasSuffix(".local") { return .bonjour }
        return .lan
    }

    /// The candidates a credential may walk automatically, in caller order.
    ///
    /// When the preferred route is protected, every cleartext candidate is
    /// removed. When it is local, that *one exact route* is the user's explicit
    /// choice and may be tried, followed only by routes which strengthen the
    /// transport. Other LAN/Bonjour addresses remain stored for display and a
    /// future manual choice, but never receive a token speculatively.
    public static func automaticCandidates(
        from candidates: [CompanionEndpoint]
    ) -> [CompanionEndpoint] {
        guard let preferred = candidates.first else { return [] }
        var seen = Set<String>()
        return candidates.filter { candidate in
            guard seen.insert(candidate.url).inserted else { return false }
            if preferred.protectsCredentials { return candidate.protectsCredentials }
            return candidate.url == preferred.url || candidate.protectsCredentials
        }
    }

    private static func normalizedURL(_ raw: String, kind: CompanionEndpointKind) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.utf8.count <= 2_048,
              var components = URLComponents(string: trimmed),
              let scheme = components.scheme?.lowercased(),
              scheme == "http" || scheme == "https",
              let host = components.host,
              !host.isEmpty,
              components.user == nil,
              components.password == nil,
              components.query == nil,
              components.fragment == nil,
              components.path.isEmpty || components.path == "/"
        else { return nil }

        switch kind {
        case .hosted:
            guard scheme == "https" else { return nil }
        case .tailnet:
            guard scheme == "http", validTailnetHost(canonicalDNSHost(host)) else { return nil }
        case .lan, .bonjour:
            guard scheme == "http" else { return nil }
        }

        components.scheme = scheme
        components.host = host.lowercased()
        components.path = ""
        if let port = components.port, !(1...65_535).contains(port) { return nil }
        return components.url?.absoluteString
    }

    private static func canonicalDNSHost(_ host: String) -> String {
        var canonical = host.lowercased()
        if canonical.hasPrefix("[") && canonical.hasSuffix("]") {
            canonical = String(canonical.dropFirst().dropLast())
        }
        while canonical.hasSuffix(".") { canonical.removeLast() }
        return canonical
    }

    private static func validTailnetHost(_ host: String) -> Bool {
        host.count > ".ts.net".count && host.hasSuffix(".ts.net")
    }

    private enum CodingKeys: String, CodingKey { case url, kind, priority }

    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        let url = try values.decode(String.self, forKey: .url)
        let kind = try values.decode(CompanionEndpointKind.self, forKey: .kind)
        let priority = try values.decode(Int.self, forKey: .priority)
        guard let accepted = CompanionEndpoint(url: url, kind: kind, priority: priority) else {
            throw DecodingError.dataCorruptedError(
                forKey: .url,
                in: values,
                debugDescription: "Companion endpoints must be absolute authorities; hosted routes require HTTPS and tailnet routes require an HTTP .ts.net name."
            )
        }
        self = accepted
    }
}

extension Connection {
    public var displayAddress: String {
        activeEndpoint?.displayAddress ?? "\(host):\(port)"
    }

    /// The normalized network origin a person must consent to before pairing.
    /// It deliberately contains no path, query, pairing code, or credential.
    public var pairingConsentOrigin: String {
        if let activeEndpoint { return activeEndpoint.url }

        var components = URLComponents()
        components.scheme = "http"
        components.host = Self.urlHost(host.lowercased())
        components.port = port
        return components.url?.absoluteString ?? "\(host.lowercased()):\(port)"
    }
}
